### PR TITLE
Add hypothesis tests for data operators.

### PR DESCRIPTION
**Pull request by hodgestar**
**Description**
Add hypothesis strategies for data objects and some simple property-based tests for data operations.

**Related issues or PRs**
- None

### CODE REVIEW

**Comment by hodgestar** [comment]
@Zac-HD Any suggestions for ways we could use hypothesis better? It all looks quite clean right now.

Is there a way we can change what hypothesis prints when an example is found? E.g. we don't want the data layer object repr to include the contents of the array, but it would be nice to see that in the test failure output.

File: qutip/tests/core/data/test_operators_hypothesis.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        result.to_array(), a.to_array() / x, rtol=1e-15
+    )
+
+
+@given(qst.qobj_datas(shape=same_shape), qst.qobj_datas(shape=same_shape))
+def test_data_equality_operator(a, b):
+    result = (a == b)
+    assert result == numpy.allclose(
+        a.to_array(), b.to_array(), rtol=1e-15
+    )
```
Comment (Zac-HD) [inline qutip/tests/core/data/test_operators_hypothesis.py:83]: This looks like it should use exact equality rather than allclose?  I'd also add a test that these objects are always equal to themselves.

File: qutip/tests/core/data/test_operators_hypothesis.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+
+same_shape = st.shared(qst.qobj_shapes())
+
+@given(qst.qobj_datas(shape=same_shape), qst.qobj_datas(shape=same_shape))
+def test_data_add_operator(a, b):
+    result = a + b
+    numpy.testing.assert_allclose(
+        result.to_array(), a.to_array() + b.to_array(), rtol=1e-15
+    )
```
Comment (Zac-HD) [inline qutip/tests/core/data/test_operators_hypothesis.py:36]: You *could* use a `@pytest.mark.parametrize("op", [operator.add, ...])` here, but it's really a matter of taste.

**Review by Zac-HD — COMMENTED** [review]
All looks good to me!

> Is there a way we can change what hypothesis prints when an example is found?

You can't change the failing example (except by changing the repr or hacking our internal pretty-printer - please don't!), but you *can* add a call to `hypothesis.note("whatever representation you want to see")`  inside the test function, and you'll see that on failure.

File: qutip/tests/core/data/test_operators_hypothesis.py (hunk 1 above)
Comment (Ericgig) [reply qutip/tests/core/data/test_operators_hypothesis.py:83]: The `Data.__eq__` test up to the numerical tolerance in core's settings which is `1e12` per default. So this allclose is stricter than the operation.

File: qutip/tests/core/data/test_operators_hypothesis.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        result.to_array(), a.to_array() - b.to_array(), rtol=1e-15
+    )
+
+
+@given(qst.qobj_datas(shape=same_shape), qst.qobj_datas(shape=same_shape))
+def test_data_matmul_operator(a, b):
+    result = a @ b
+    numpy.testing.assert_allclose(
+        result.to_array(), a.to_array() @ b.to_array(), rtol=1e-15
+    )
```
Comment (AGaliciaMartinez) [inline qutip/tests/core/data/test_operators_hypothesis.py:52]: For matrix multiplication the following restriction would be required `a.shape[1]==b.shape[0]`. Requiring same shape will lead to errors for non square matrices. I wonder if there is a quick way of generating such strategy similar to how same_shape works.

**Review by AGaliciaMartinez — COMMENTED** [review]
The hypothesis testing approach seems to generate very clean and readable code! I really like this approach for testing as seems to simplify writing tests. I see that some tests do not fully work yet but the errors seem to be mostly related to inf and nan values not being properly handled. We can skip those tests by using the elements argument in `npst.arrays` but im not sure if we should. Also, absolute tolerances may need to be included in tests if we test with values close to zero.

@hodgestar What is the goal for this PR? Is this meant to ultimately replace the current testing code employed for the data layer? or is it just mean to test Data's operators?

File: qutip/tests/core/data/test_operators_hypothesis.py (hunk 3 above)
Comment (Zac-HD) [reply qutip/tests/core/data/test_operators_hypothesis.py:52]: Check out [the `mutually_broadcastable_shapes()` strategy](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.mutually_broadcastable_shapes); you can see a quick example from `$ hypothesis write numpy.matmul` in your terminal 🙂

**Comment by Zac-HD** [comment]
(unsubscribing because I think there's not much for me to add, but please @ me if that changes)

File: qutip/tests/core/data/test_operators_hypothesis.py (hunk 2 above)
Comment (hodgestar) [reply qutip/tests/core/data/test_operators_hypothesis.py]: I quite like the separate tests.

File: qutip/tests/core/data/test_operators_hypothesis.py (hunk 3 above)
Comment (hodgestar) [reply qutip/tests/core/data/test_operators_hypothesis.py]: I ended up writing my own simpler `MatrixShapesStrategy` because I couldn't get `MutuallyBroadcastableShapes` to not generate any extra loop dimensions.

I also added a helper `qobj_shared_shapes` that allows `shape_a, shape_b = qobj_shared_shapes([("m", "k"), ("k", "n")])` so that for tests one can write `@given(shape_a, shape_b)` and get the compatible shapes.

**Comment by hodgestar** [comment]
@Ericgig @AGaliciaMartinez This is ready for another round of review.

I fixed small bugs in `mul` and `iszero` -- we should discuss whether the fixes are good or not.

There are still some failing examples I'm encountering locally, that we should also decide what to do with:

**scalar division**:
```
# result = a / x
# numpy returns array([[nan+nanj]])
# our result:
result [Data, Dense]: array([[8.98846567e+307+8.98846567e+307j]])
a [Data, Dense]: array([[9.97920155e+291+1.79769313e+308j]])
x: (1+1j)
```

**matmul**:
The current failing example I'm seeing locally is just an `atol` issue encountered for large exponents, but I have encounter other failing examples still that I need to dig into.

**Comment by hodgestar** [comment]
@Zac-HD Would you mind doing a quick sanity check of my `MatrixShapesStrategy` class (https://github.com/qutip/qutip/blob/2aee70cb3981e60e236de831c56505a7caabd0aa/qutip/tests/strategies.py#L15-L62)? Mostly I want to know whether I should have been able to do this with `mutually_broadcastable_shapes`.

I also wouldn't mind a quick check of `qobj_shared_shapes` (https://github.com/qutip/qutip/blob/2aee70cb3981e60e236de831c56505a7caabd0aa/qutip/tests/strategies.py#L113-L131). Is this a good thing to be doing with shared strategies?

**Comment by Zac-HD** [comment]
> Mostly I want to know whether I should have been able to do this with `mutually_broadcastable_shapes`.

Yes, you should use [`mutually_broadcastable_shapes(signature=)`](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.mutually_broadcastable_shapes) for that.  The signature syntax is a little different, since Numpy does it with strings, but we already support named dimensions, constant-size dimensions, and optional dims.  Should be a bit more efficient too.

> I also wouldn't mind a quick check of `qobj_shared_shapes`. Is this a good thing to be doing with shared strategies?

It does technically work?  But in such a situation I'd probably just use `st.data()` and draw in the body of the test; see `hypothesis write numpy.matmul` for an example.  If you expected to use it in many tests maybe a custom strategy, but for just `test_data_matmul_operator` that doesn't seem worth it.

**Comment by hodgestar** [comment]
> Yes, you should use [`mutually_broadcastable_shapes(signature=)`](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.mutually_broadcastable_shapes) for that. The signature syntax is a little different, since Numpy does it with strings, but we already support named dimensions, constant-size dimensions, and optional dims. Should be a bit more efficient too.

I tried with `signature=` quite a bit but `(m, k), (k, n)` always produced thing like `(1, 1, 1), (1, 1)` despite setting `max_dims` and `min_dims` to various things.
 
> It does technically work? But in such a situation I'd probably just use `st.data()` and draw in the body of the test; see `hypothesis write numpy.matmul` for an example. If you expected to use it in many tests maybe a custom strategy, but for just `test_data_matmul_operator` that doesn't seem worth it.

I expected there to be other use cases.  I started with the output of `write numpy.matmul` but ended up playing around with my own strategy (see difficulties with mutually_broadcastable_shapes` above) and liking the idea of not have to unpack `a` and `b` inside the test, and of being able to supply the shapes for `a` and `b` to other strategies (e.g. maybe we will have a strategy for Qobj and QobjEvo objects and want to try various combinations of multiplying things together)

**Comment by Zac-HD** [comment]
> I tried with `signature=` quite a bit but `(m, k), (k, n)` always produced thing like `(1, 1, 1), (1, 1)` despite setting `max_dims` and `min_dims` to various things.

Huh, that's surprising.  Share some example code and I'll have a look?

> I expected there to be other use cases. I started with the output of `write numpy.matmul` but ended up playing around with my own strategy (see difficulties with mutually_broadcastable_shapes`above) and liking the idea of not have to unpack`a`and`b`inside the test, and of being able to supply the shapes for`a`and`b` to other strategies (e.g. maybe we will have a strategy for Qobj and QobjEvo objects and want to try various combinations of multiplying things together)

👍 sounds good, carry on then!

**Comment by hodgestar** [comment]
@Ericgig and I had a brief discussion about how to handle `inf` and `nan` generally. One option is forbid them entirely from the data layer -- i.e. to check and raise an error whenever a new data layer is created. A potential downside to this is we'd have to check often to be sure, since any operation can potentially create `nan` and `inf`, even simple addition:
```python
>>> 1e308 + 1e308
inf
>>> (1e308 + 1e308) - (1e308 + 1e308)
nan
```

**Comment by Zac-HD** [comment]
(out of my wheelhouse again; @ me again if that changes (again!))

**Review by Ericgig — COMMENTED** [review]
Hypothesis still feel somewhat magic, but I am starting to understand it.
Right now it seems to create data with mostly full arrays. I don't know how hard it is to create a strategy for different densities and to add the flags for Hermitian, etc. but it will be needed eventually.

The tolerances used are too tight, making some tests fail.


Also, from the old scipy test:
```
>>> a = np.array([[7.59109827e+14+9.96089747e+14j]])
>>> b = np.array([[7.22335902e+307+1.64618697e+308j]])
>>> np.allclose(b, a), np.allclose(a, b)
(False, True)
```
I am not sure we should ignore the overflow warning...

File: qutip/tests/strategies.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        min_side=1,
+        max_side=64,
+    ):
+        super().__init__()
+        self.side_strat = st.integers(min_side, max_side)
+        self.shapes = tuple(shapes)
+        self.min_side = min_side
+        self.max_side = max_side
+
+    def do_draw(self, data):
```
[the remark this answers is not in the document]
Comment (hodgestar) [reply qutip/tests/strategies.py:47]: The `data` is the thing that manages drawing samples from strategies (via `data.draw(strategy)`). I don't know what it looks like under the hood, but I imagine it abstractly as something like a tree of counters that help draw the right examples for the sub-strategies within more complex strategies like this one.

**Comment by hodgestar** [comment]
@Ericgig I've expanded the tests and reworked this so that it's less of a WIP. I've also documented the expected behaviour of the data layer in the presence of `nans`, etc in the description of the PR. I'll find somewhere in the docs to add that before merging the PR.

Would you mind taking another look and letting me know what you think?

File: qutip/core/data/mul.pyx [hunk 5]
```diff
@@ -15,15 +17,24 @@ cpdef CSR imul_csr(CSR matrix, double complex value):
...
+            f"Multiplying CSR matrix by non-finite value {value!r}"
+            " would generate a dense matrix of NaNs."
+        )
     zscal(&l, &value, matrix.data, &ONE)
     return matrix
 
 cpdef CSR mul_csr(CSR matrix, double complex value):
     """Multiply this CSR `matrix` by a complex scalar `value`."""
-    if value == 0:
-        return csr.zeros(matrix.shape[0], matrix.shape[1])
```
Comment (Ericgig) [inline qutip/core/data/mul.pyx:24]: I hate the removal of this shortcut.
`mul(data, 0)` can happen often in QobjEvo when the coefficient is a pulse or step function.
Without this shortcut, they need to loop over the values twice, once for this `mul`, once for the `add` or `matmul` following.
Thus this shortcut will actually help users.
It is also one of the only places where there is a `mul(data, 0)` in the code.

In some theoretical case, it can shallow a NaN, but I don't see any realistic case where this could hide an error.

So everyone loses for a case that will probably never happen...

Can you write a code example where this would save us? Something that looks like it should work, but cause infinities or NaN that would be erased by a multiplication by 0 without returning garbage.

If we can't even think of such a case, let's keep the shortcut.
If you can find one, let's loop over the values to find Nans and return the zeros crs if none are found so we at least loop only once over the values.

Comment (hodgestar) [reply qutip/core/data/mul.pyx:24]: Maybe we just need to make `_data.zeros[type(x)](*x.shape)` easier to use so that we don't need to write `mul(data, 0)`? I can add `_data.zeros_like(x)` if that would be better?

In the specific case where `mul(data, 0)` is intended to create a matrix of zeros of the right shape, the current fast path can never be wrong since that is exactly what it does.  In general, `mul(data, 0)` is currently just incorrect floating point arithmetic.

I was wondering about adding `has_non_finite` flag to CSRs. It will be a bit of pain to keep track of everywhere but it would allow us to fast path the all finite cases without sacrificing correctness.

If this is a blocker for this PR, I'm happy to undo the current change and just make the tests pass by just explicitly stating that the expected result is `0` in those cases in the test.

I also don't want to sacrifice performance, but it is important to have a clear set of tests that show how we expect the data layer to behave in these edge cases.

Comment (Ericgig) [reply qutip/core/data/mul.pyx:24]: I am more thinking about the use in `QobjEvo`. I don't think we ever use it as `_data.zeros_like`, in the code base, but that function would be useful, as would be an `identity_like` as we discussed somewhere.

I would be fine with a `has_non_finite` flag, but I think we should raise an error anytime we would raise that flag.
I think that we could do the check when we tidyup: we already check the scale of all values and it's done in most operation that could create zeros, thus functions at risk.

I see a CSR matrix full of zeros as wrong, if not an error, so this is a blocker to me. 
This and tests should pass reliably.

Comment (Ericgig) [reply qutip/core/data/mul.pyx:24]: I am using `mul` as `_data.zeros_like` in stochastic...

**Comment by hodgestar** [comment]
Closing -- I don't have bandwidth for this currently. Let's create a new MR when we have capacity again.
